rootfs/api/tests/test_build.py

```diff
@@ -235,6 +236,144 @@ def test_build_default_containers(self, mock_requests):
+        url = "/v2/apps/{app_id}/builds".format(**locals())
+        body = {'image': 'autotest/example'}
+        response = self.client.post(url, body)
+        self.assertEqual(response.status_code, 201, response.data)
+
+        # verify worker is still there
+        url = "/v2/apps/{app_id}/pods/worker".format(**locals())
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200, response.data)
+        self.assertEqual(len(response.data['results']), 0)
+
+        # verify web is still there
+        url = "/v2/apps/{app_id}/pods/web".format(**locals())
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200, response.data)
+        self.assertEqual(len(response.data['results']), 0)
+
+        # look at the app structure
+        url = "/v2/apps/{app_id}".format(**locals())
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200, response.data)
+        self.assertEqual(response.json()['structure'], {'web': 0, 'worker': 0})
+
+    @override_settings(PROCFILE_REMOVE_PROCS_ON_DEPLOY="")
+    def test_build_no_remove_process(self, mock_requests):
+        """
+        Specifically test PROCFILE_REMOVE_PROCS_ON_DEPLOY being turned off
+        and a user posting a new build without a Procfile that was previously
+        applid
+        """
+        # start with a new app
+        app_id = self.create_app()
+
+        # post a new build with procfile
+        url = "/v2/apps/{app_id}/builds".format(**locals())
+        body = {
+            'image': 'autotest/example',
+            'sha': 'a'*40,
+            'procfile': json.dumps({
+                'web': 'node server.js',
+                'worker': 'node worker.js'
+            })
+        }
+        response = self.client.post(url, body)
+        self.assertEqual(response.status_code, 201, response.data)
+
+        # scale worker
+        url = "/v2/apps/{app_id}/scale".format(**locals())
+        body = {'worker': 1}
+        response = self.client.post(url, body)
+        self.assertEqual(response.status_code, 204, response.data)
+
+        # verify worker
+        url = "/v2/apps/{app_id}/pods/worker".format(**locals())
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200, response.data)
+        self.assertEqual(len(response.data['results']), 1)
+        container = response.data['results'][0]
+        self.assertEqual(container['type'], 'worker')
+        self.assertEqual(container['release'], 'v2')
+        # pod name is auto generated so use regex
+        self.assertRegex(container['name'], app_id + '-worker-[0-9]{8,10}-[a-z0-9]{5}')
+
+        # do another deploy for this time forget Procfile
+        url = "/v2/apps/{app_id}/builds".format(**locals())
+        body = {'image': 'autotest/example'}
+        response = self.client.post(url, body)
+        self.assertEqual(response.status_code, 201, response.data)
```

Comment:
> I could try to look at doing this in such a way where the flag only matters if you previously had a Procfile but none is submitted now. 

@helgi, what would it look like then?

Lets say I have this app (deploy was made with deis pull and a Procfile specified):

```
=== xapp Processes
--- cmd:
xapp-cmd-1286488693-43lgu up (v92)
--- worker:
xapp-worker-1872629609-70hu1 up (v92)
```

Later I make a deis pull without specifying a Procfile, then, deis would reject the pull with something like:

```
This deploy is not allowed because it would delete process types and you have the PROCFILE_REMOVE_PROCS_ON_DEPLOY flag set to false.

Processes before:
--- cmd:
xapp-cmd-1286488693-43lgu up (v92)
--- worker:
xapp-worker-1872629609-70hu1 up (v92)

Processes after:
--- cmd:
xapp-cmd-1286488693-43lgu up (v93)

If you really want to make the change, use the -x flag (x can be -f or -y, don't know...)

```

What do you think @helgi? @heynemann any thoughts here?